test/strings/basic.jl

```diff
@@ -428,6 +428,9 @@ foobaz(ch) = reinterpret(Char, typemax(UInt32))
 # issue #17624, missing getindex method for String
 @test "abc"[:] == "abc"
 
-# PR #18259 reverted because failing test below
-a = [x for x in String([0xcf, 0x83, 0x83, 0x83, 0x83])]
-@test a[1] == 'σ'
+# we assume for performance that next/nextind return values past the end of a
+# string's underlying data; this helps with performance of `done`.
```

Comment:
Let me clarify the changes in behaviour versus the previous behaviour with a few examples:

Say we have a string, "S", and append to it some continuation characters.

```julia
julia> s = string("S", String([0x83, 0x83, 0x83, 0x83]));
```

In both previous and current behaviour, `start` will return 1, and `done` and `next` will give the expected result:

```julia
julia> start(s)
1

julia> done(s, 1)
false

julia> next(s, 1)
('S',2)
```

In the previous version, `done` will now return `true`, as the remaining data in the string is composed entirely of continuation characters, which `endof` skips. Whereas now, `done` will return `false`, as it believes that the character at `2` should be a valid index. `next` will then throw up on the index that isn't right.

This leads to the observed behaviour change:

- In the previous version, `collect(string("S", String([0x83, 0x83, 0x83, 0x83])))` returned just `['S']`, discarding junk at the end of the string.
- In the current version, the same will fail.

It is not clear to me which is better, and I suspect whether silently pretending bad data does not exist versus raising an error on encountering it are both undesirable in their own ways. Note that both previous and current version throw an exception on `collect(string("S", String([0x83, 0x83, 0x83, 0x83]), "S"))`; the fact that only continuation characters end a string is the only reason the previous version was able to `collect` this case.

----

There is, to my knowledge, just one other case where the behaviour changes, and that is when `endof` itself throws an exception. Thus in the previous version, a `BoundsError` is generated by `endof` on a string consisting only of continuation characters; in the new version, a `UnicodeError` is generated by `next`.

----

Anyhow, I really don't think any of these behaviours are documented, reliable, or should factor strongly. I personally prefer the new behaviour because it is more consistent with concatenation—the new behaviour guarantees that two strings that collect properly will concatenate into a string that collects properly; the old behaviour makes no such guarantee. But I really think performance considerations should take priority over how exactly invalid data is handled.